sycl/unittests/helpers/PiMock.hpp

```diff
@@ -0,0 +1,193 @@
+    // Copy the PiPlugin, thus untying our to-be mock platform from other
+    // platforms within the context. Reset our platform to use the new plugin.
+    auto NewPluginPtr = std::make_shared<detail::plugin>(OriginalPiPlugin);
+    ImplPtr->setPlugin(NewPluginPtr);
+    // Extract the new PiPlugin instance by a non-const pointer,
+    // explicitly allowing modification
+    auto &PiPluginRef = const_cast<RT::PiPlugin &>(NewPluginPtr->getPiPlugin());
+    MPiPluginMockPtr = &PiPluginRef;
+    // Save a copy of the platform resource
+    MPlatform = OriginalPlatform;
+  }
+
+  /// Constructs PiMock from a device_selector, provided that
+  /// a non-host device can and will be selected.
+  ///
+  /// \param DevSelector is a reference to a device_selector instance.
+  explicit PiMock(const cl::sycl::device_selector &DevSelector)
+      : PiMock(cl::sycl::platform{DevSelector}) {}
+
+  /// Explicit construction from a host_selector is forbidden.
+  PiMock(const cl::sycl::host_selector &HostSelector) = delete;
+
+  PiMock(const PiMock &) = delete;
+  PiMock &operator=(const PiMock &) = delete;
+  ~PiMock() = default;
+
+  /// Returns a handle to the SYCL platform instance.
+  ///
+  /// \return A reference to the SYCL platform.
+  cl::sycl::platform &getPlatform() { return MPlatform; }
+
+  template <detail::PiApiKind PiApiOffset>
+  using FuncPtrT = typename RT::PiFuncInfo<PiApiOffset>::FuncPtrT;
+  template <detail::PiApiKind PiApiOffset>
+  using SignatureT = typename std::remove_pointer<FuncPtrT<PiApiOffset>>::type;
+
+  /// Redefines the implementation of a given PI API to the input
+  /// function object.
+  ///
+  /// \param Replacement is a mock std::function instance to be
+  ///        called instead of the given PI API. This function must
+  ///        not have been constructed from a lambda.
+  template <detail::PiApiKind PiApiOffset>
+  void redefine(const std::function<SignatureT<PiApiOffset>> &Replacement) {
+    // TODO: Find a way to store FPointer first so that real PI functions can
+    // be called alongside the mock ones. Something like:
+    // `enum class MockPIPolicy { InsteadOf, Before, After};`
+    // may need to be introduced.
+    FuncPtrT<PiApiOffset> FuncPtr =
+        *Replacement.template target<FuncPtrT<PiApiOffset>>();
+    assert(FuncPtr &&
+           "Function target is empty, try passing a lambda directly");
+    setFuncPtr<PiApiOffset>(MPiPluginMockPtr, *FuncPtr);
+  }
+
+  /// A `redefine` overload for function pointer/captureless lambda
+  /// arguments.
+  ///
+  /// \param Replacement is a mock callable assignable to a function
+  ///        pointer (function pointer/captureless lambda).
+  template <detail::PiApiKind PiApiOffset, typename FunctorT>
+  void redefine(const FunctorT &Replacement) {
+    // TODO: Check for matching signatures/assignability
+    setFuncPtr<PiApiOffset>(MPiPluginMockPtr, Replacement);
+  }
+
+private:
+  cl::sycl::platform MPlatform;
```

Comment:
[This comment](https://github.com/intel/llvm/pull/1534#discussion_r410035980), and the thread in general elaborate as to why wrapping a SYCL object might be needed at all. A brief summary is, the user code might want to pass a SYCL object to some function, but also pass a handle to modify the PI calls. Having something that combines both is convenient for the unittest code.

Inheriting from `detail::plugin` actually was my initial idea. The problem arose, however: to reach the `redefine`-like API from the user code...
1. either the user code would have to extract this `plugin` derivative from a SYCL object;
2. or the higher-level SYCL objects would also have to be extended with `redefine` (calling `MockPlugin.redefine()` under the hood).

**1** would be fine, I guess, but could result in weird code when constructing from high-level SYCL object - having to redirect their `detail::plugin` handle to `this`... Probably just a matter of taste :D Additionally, the problem of passing around the to-be-mocked SYCL objects from the first paragraph would persist. After receiving this SYCL object into some function, not only would the `detail::plugin` object have to be reached, but also a `static_cast` to the hypothetical `MockPlugin` would be required.

**2** would also have to be achieved by inheritance from all relevant `*_impl` objects, which would arguably result in less understandable code (e.g. to add creation from `queue`, we'd also have to extend a `queue_impl`, instead of just implementing a 2-line delegating constructor).

I might be missing some other approach, though, so I can rework this in some cleaner way that also retains the usability. What do you think here?

Reply:
Would something like this work ?
~~~
auto & plugin = platform.getPlugin();
PiMock mock(plugin);
mock.redefine(...)
platform.setPlugin(mock);
~~~

where the `getPlugin/setPlugin` work with `detail::plugin` and `PiMock` is inherited from `detail::plugin` and extends it with the ability to redefine some entries in the table of APIs.
And of course, the plugin is accessible through every SYCL object, not just the platform.

Reply:
1. The `PiPlugin` field of `detail::plugin` should be made `protected`.
2. This user code would have to look like:
```
auto &plugin = getSyclObjImpl(platform)->getPlugin();
PiMock mock(plugin);
mock.redefine<...>(...);
getSyclObjImpl(platform)->setPlugin(PiMock);
```
and `plugin` extraction would have to differ for various SYCL objects.

To avoid such verbose user code, we'd have to encapsulate this like:
```
PiMock(const cl::sycl::platform &OriginalPlatform) : 
  PiMock(getSyclObjImpl(OriginalPlatform)->getPlugin()) {
  getSyclObjImpl(OriginalPlatform)->setPlugin(this);
}
```
This looks quite decent to me, actually - an API like `getPlatform()` could still be provided by simply building a platform object out of `this`.

Reply:
Looks good to me, and then maybe call `PiMock` a `MockPlugin` since this would be just a plugin.

Reply:
This runs into a double `delete` problem.
```c++
{
  platform P;
  MockPlugin Mock(P);
  // The previous line under the hood:
  MockPlugin(platform &P) : detail::plugin(getSyclObjImpl(P)->getPlugin()) {
    // Transfer plugin's ownership to the impl object
    getSyclObjImpl(P)->setPlugin(std::make_shared<plugin>(*this);
  }
  // Back to the original scope
} // End of scope; the stack-allocated Mock is deleted.
  // The platform is also destructed, and a second delete
  // is attempted on the plugin that is supposedly owned by this platform.
```
It's actually funny how the current implementation's storage of a `platform` ensures that `shared_ptr<detail::plugin>`'s refcount is always nullified at the same time that a `PiMock` instance goes out of scope. I must confess I haven't thought about it :)

Given the current findings, is there a way to proceed with such a refactoring and not make things even more obscure?

Reply:
I think we should restore the real plugin upon destruction of the mock. It will solve the double delete of the mock, and ensure that the original plugin is destroyed too.

Reply:
Sounds like a plan - I'll try this, thanks.

Reply:
Sorry this took a lot of time.

The issue with restoring the plugin is that during `MockPlugin`'s destruction, `this` is already invalidated, and since it's pointed to by `platform_impl::MPlugin` instances, applying `MPlugin.reset(RestoredPluginObj)` (or `swap`) [crashes](https://en.cppreference.com/w/cpp/memory/shared_ptr/reset) the process.

Consequently, exposing `MockPlugin`'s allocation to the user stack ins't feasible. I could still try reworking this into a helper function that returns a reference to the `platform_impl`-owned `MockPlugin`, or retain the current approach.

Reply:
Which "**this** is invalidated"? It should be perfectly valid to still access this in a destructor, I believe. 

It is hard to design things in the conversations on github, so since it does not come easy enough, and since it is entirely in the unittest and does not influence the rest of the SYCL design, I am fine with the approach that you have working already (with maybe a TODO to re-design if wanted).

Reply:
`this` was probably a wrong term to use - what I've meant was "the `plugin *` pointer to the instance of `MockPlugin` which is being destructed".
Then I'll add a `TODO` and address @keryell's comment.